---
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: spacelab           # flatly spacelab sandstone cerulean
    code_download: true
title: "`r paste('Transcriptome analysis:', gsub(pattern = '_', replacement = ' ', x = params$project_name))`"
subtitle: "`r paste(params$platform, 'analysis', ifelse(!is.null(params$group_filter), paste('for', paste(params$group_filter, collapse=' and ')), ''))`"
---


***

# Revision History

**This version**  
  
*Current version: 1*  
Date report generated: `r format(Sys.time(), '%d %B, %Y')`  
Report prepared for: Name  
Purpose of report:  

* Exercise to analyze RNA-Seq data  

**Previous revisions**  

N/A  

***


# TODO: Rmd parameters like author and prepared for should be in the config

This report is meant to help explore DESeq2 results and was generated using RMarkdown. This section contains the code for setting up the rest of the report.  

## Load libraries

```{r docSetup, warning = FALSE, message = FALSE}
library('tidyverse')
library('knitr')
library('kableExtra')


```

## load functions
```{r load_functions}



```

## load config
```{r load_config}

config <- yaml::read_yaml(file.path(here::here(),
                                    "config/config.yaml"),
                          eval.expr = T)


```


```{r load_data, collapse=TRUE}

load(params$dataFile) # metadata, contrasts, counts, resultsList



```


# metadata tables
```{r tables, collapse=TRUE}

# table of sample metadata
metadata_table <- metadata %>% arrange(!!sym(design_to_use))
knitr::kable(metadata_table,
             row.names =  F,
             caption = "Sample metadata") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")

# table of contrasts
knitr::kable(contrasts,
             row.names =  F,
             caption = "Sample metadata") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")

# table of #samples/group
nSamples_table <- table(metadata_table %>% dplyr::pull(!!sym(params$design))) 
knitr::kable(nSamples_table,
             row.names =  F,
             caption = "Sample metadata") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")

# table of #DEGs/contrast 
summary_counts <- data.frame()
comparisons <- names(resultsList)
for(comp in comparisons){ # by comparison
    res <- resultsList[[comp]]
    c <- nrow(res)
    row <- data.frame(comparison=comp, DEG=c)
    summary_counts <- rbind(summary_counts, row)
}

knitr::kable(summary_counts,
             row.names =  F,
             caption = "Sample metadata") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")



```


# descriptive plots

# boxplots of reads/sample (by group)
```{r nReads_boxplot, collapse=TRUE}
df<-data.frame(colSums(counts))
names(df) <- c("nReads")
df$original_names = rownames(df)
df2 <- df %>%
    left_join(metadata) %>%
    dplyr::select(original_names,sym(design_to_use),nReads)

ggplot(df2, aes_string(x=design_to_use, y="nReads") )+
    geom_boxplot() +
    xlab("Group") +
    ylab("Number of reads") +
    theme_bw()

```

# PCA plots (all genes, DEGs only, top 1000 most variable genes)
```{r PCA, collapse=TRUE}

plotPCA(rld, intgroup = params$design, ntop = nrow(assay(rld))) + 
  theme(legend.position = "bottom") +
  guides(color = guide_legend(ncol = 3))

```

# sample-to-sample heatmaps
# DEG heatmaps

# Differential gene expression plots

# MA plots
# volcano plots
# p-value distribution
# DEG gene info table
# gene-level plots across timepoints (?)

# Pathway analysis plots

# todo

# methods summary
# references
