---
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: spacelab           # flatly spacelab sandstone cerulean
    code_download: true
---




***

# Revision History

**This version**  
  
*Current version: 1*  
Date report generated: `r format(Sys.time(), '%d %B, %Y')`  
Report prepared for: Name  
Purpose of report:  

* Exercise to analyze RNA-Seq data  

**Previous revisions**  

N/A  

***



```{r docSetup, warning = FALSE, message = FALSE}
library('tidyverse')
library('knitr')
library('kableExtra')
library('DESeq2')
library('RColorBrewer')
library('pheatmap')
library('data.table')


source(here::here("scripts","file_functions.R"))
source(here::here("scripts","setup_functions.R"))
source(here::here("scripts","DESeq_functions.R"))
source(here::here("scripts","data_functions.R"))

#########################################
# to delete, should all be done by the render_DEseq2_reports.R script. Just in here to help with development


config <- yaml::read_yaml(file.path(here::here(),
                                    "config/config.yaml"),
                          eval.expr = T)

# Combine required params from config
params <- c(config$common, config$DESeq2)
# replace nulls in params with NA
params <- replace_nulls_in_config(params)
# If projectdir is not set, figure out current project root directory
projectdir <- params$projectdir
if (is.na(projectdir)) {
  projectdir <- here::here()
  params$projectdir <- projectdir
}

paths <- set_up_paths(params)


skip_extra <- c("DMSO") # Remove DMSO controls as a facet

# Input file - Rmd
inputFile <- file.path(projectdir, "Rmd", "DESeq2_report_new.Rmd")

# Identify where metadata can be found
SampleKeyFile <- file.path(projectdir, "data/metadata/metadata.QC_applied.txt")

params$dataFile <- file.path(paths$DEG_output, paste0(params$project_name, "_DEG_data.RData"))

#########################################

species_data <- load_species(params$species)

bs <- load_biospyder_new(params$biospyder_dbs, species_data$temposeq_manifest)


```


## load config
```{r load_config}



```


```{r load_data}
load(params$dataFile) # metadata, contrasts, counts, resultsList

# reformat data based on group_facet and display_group_facet

# case 1: no facet, no display facet
if(is.na(params$group_facet) && is.na(params$display_group_facet)){
    dds <- ddsList[['all']]
    resultsListAll <- overallResListAll[['all']]
    resultsListDEGs <- overallResListDEGs[['all']]
    rld <- rldList[['all']]
    mergedDEGs <- mergedDEGsList[['all']]

# case 2: no facet, yes display facet
} else if(is.na(params$group_facet) && !is.na(params$display_group_facet)){
  # the data isn't already faceted but we need to extract the facet we need
    display_group_filter <- params$display_group_filter
  
    dds_all <- ddsList[['all']]
    resultsListAll_all <- overallResListAll[['all']]
    resultsListDEGs_all <- overallResListDEGs[['all']]
    rld_all <- rldList[['all']]
    mergedDEGs_all <- mergedDEGsList[['all']]
    
    metadata_subset <- subset_metadata(designList[['all']], design_to_use, contrasts, params$display_group_facet, display_group_filter)
    DESeqDesign_subset <- metadata_subset$DESeqDesign
    contrasts_subset <- metadata_subset$contrasts
    
    dds_subset <- subset_data(dds_all, DESeqDesign_subset)
    rld_subset <- subset_data(rld_all, DESeqDesign_subset)
    
    contrast_strings <- contrasts_subset %>% mutate(contrast_string = paste(V1,V2,sep="_vs_")) %>% pull(contrast_string)
    resultsListAll_subset <- resultsListAll_all[contrast_strings]
    resultsListDEGs_subset <- resultsListDEGs_all[contrast_strings]

    # note, in this case the merged DEGs will be for the whole experiment, not the display facet
    DESeqDesign <- DESeqDesign_subset
    contrasts <- contrasts_subset
    dds <- dds_subset
    resultsListAll <- resultsListAll_subset
    resultsListDEGs <- resultsListDEGs_subset
    rld <- rld_subset
    mergedDEGs <- mergedDEGs_all
    
    # TODO: add some tests here to make sure everything worked properly
    
# case 3: yes facet, yes display facet
} else if(!is.na(params$group_facet) && !is.na(params$display_group_facet)){
    display_group_filter <- params$display_group_filter
    
    dds <- ddsList[[display_group_filter]]
    resultsListAll <- overallResListAll[[display_group_filter]]
    resultsListDEGs <- overallResListDEGs[[display_group_filter]]
    rld <- rldList[[display_group_filter]]
    mergedDEGs <- mergedDEGsList[[display_group_filter]]
  
# case 4: yes facet, no display facet, this one doesn't make sense
} else {
  stop("Making a single report for faceted data not supported. Did you forget to set display_group_facet?")
}


# filter the regularized data a couple ways for different displays    
rld_DEGs <- rld[row.names(assay(rld)) %in% mergedDEGs]

rv = rowVars(assay(rld))
select = order(rv, decreasing = TRUE)[1:params$nBest]
rld_top <- rld[select,]
```


# metadata tables
```{r tables, collapse=TRUE}

# table of sample metadata
metadata_table <- DESeqDesign %>% arrange(!!sym(params$design))
knitr::kable(metadata_table,
             row.names =  F,
             caption = "Sample metadata") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")

# table of contrasts
knitr::kable(contrasts,
             row.names =  F,
             caption = "Sample metadata") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")

# table of #samples/group
nSamples_table <- table(metadata_table %>% dplyr::pull(!!sym(params$design))) 
knitr::kable(nSamples_table,
             row.names =  F,
             caption = "Sample metadata") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")

# table of #DEGs/contrast 
summary_counts <- data.frame()
comparisons <- names(resultsListDEGs)
for(comp in comparisons){ # by comparison
    res <- resultsListDEGs[[comp]]
    c <- nrow(res)
    row <- data.frame(comparison=comp, DEG=c)
    summary_counts <- rbind(summary_counts, row)
}

knitr::kable(summary_counts,
             row.names =  F,
             caption = "Sample metadata") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")



```


# descriptive plots

# boxplots of reads/sample (by group)
```{r nReads_boxplot, collapse=TRUE}
df<-data.frame(colSums(counts(dds)))
names(df) <- c("nReads")
df$original_names = rownames(df)
df2 <- df %>%
    left_join(DESeqDesign) %>%
    dplyr::select(original_names,sym(design_to_use),nReads)

ggplot(df2, aes_string(x=design_to_use, y="nReads") )+
    geom_boxplot() +
    xlab("Group") +
    ylab("Number of reads") +
    theme_bw()

```

# sample-to-sample heatmaps
```{r sample-to-sample_heatmaps}

plot_sample_heatmap <- function(current_rld){
  ## Obtain the sample euclidean distances
  sampleDists <- stats::dist(t(assay(current_rld)))
  sampleDists <- dendextend::sort_dist_mat(sampleDists)
  sampleDistMatrix <- as.matrix(sampleDists)
  
  distances_df <- as.data.frame(colData(current_rld)[, intgroup, drop = F])
  
  ## Define colors to use for the heatmap
  colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
  
  ## Make the heatmap
  return( pheatmap(sampleDistMatrix,
           clustering_distance_rows = sampleDists,
           clustering_distance_cols = sampleDists,
           color = colors,
           annotation_col = distances_df) )
}
# all genes
plot_sample_heatmap(rld)

# DEGs only
plot_sample_heatmap(rld_DEGs)

# 100 most variable genes
plot_sample_heatmap(rld_top)
```

# P-value distributions
```{r pvalue_distribution}
allResults <- annotate_deseq_table(resultsListAll, params, bs, filter_results = F)
ggplot(data.frame(allResults), aes(x = pvalue)) +
    geom_histogram(alpha = .5, position = 'identity', bins = 50) +
    labs(title = 'Histogram of unadjusted p-values') +
    xlab('Unadjusted p-values') +
    facet_wrap( ~ contrast, ncol = 2)
```

# MA plots
```{r MA_plots}
foldchangestats <- boxplot(abs(allResults$log2FoldChange), plot = F)
ylims_for_ma_plots <- c(-foldchangestats$stats[4, 1], foldchangestats$stats[4, 1])


for (i in seq_along(resultsListAll)) {
  contrast = gsub(pattern = paste0("log2.*", params$design, "\ "),
                  replacement =  "",
                  x = resultsListAll[[i]]@elementMetadata[[2]][2])
  cat("###", contrast, "  \n\n")
  MA_title <- paste0('MA plot, alpha = ', metadata(resultsListAll[[i]])$alpha,', ', contrast)
  DESeq2::plotMA(resultsListAll[[i]],
                 alpha = metadata(resultsListAll[[i]])$alpha,
                 main = str_wrap(MA_title, width = 40),
                 ylim = ylims_for_ma_plots)
  cat('  \n\n')
}

```


